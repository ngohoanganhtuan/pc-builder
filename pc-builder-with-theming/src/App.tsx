import clsx from 'clsx';
import { useState } from 'react';
import { BrowserRouter, Routes, Route } from 'react-router-dom';

// Style
import 'index.scss';

// Components
import { ErrorBoundary, Footer, Header, StoreProvider } from '@components';

// Pages
import { CartPage, SpecBuilderPage, PCPartDetail } from '@pages';
import { ThemeProvider } from '@components/Providers/ThemeProvider/ThemeProvider';
import BackgroundFixed from '@components/BackgroundFixed/BackgroundFixed';
import Sidebar from '@components/Sidebar/Sidebar';

// Enums
import { ROUTES } from '@enums';
import VariantBox from '@components/VariantBox/VariantBox';
import ThemeSidebar from '@components/ThemeSidebar/ThemeSidebar';
import { THEMES, ThemeTypeName } from '@contexts';
import { BACKGROUND_MAPPING_WITH_THUMBNAIL, BACKGROUND_THUMBNAIL } from '@constants/imagesPath';
import { ThemeVariantList } from '@components/ThemeSidebar/ThemeVariantList';
import BackgroundVariantList from '@components/ThemeSidebar/BackgroundVariantList';

// Global state cho PC (inner component)
// Global state cho Theme (inner component)

const App = () => {
  const [theme, setTheme] = useState<ThemeTypeName>(THEMES[1] as ThemeTypeName);
  const [background, setBackground] = useState<string>('BACKGROUND_THUMB_LIGHT_6');
  const [isOpenSidebar, setToggleSidebar] = useState<boolean>(false);

  const handleChangeTheme = (theme: ThemeTypeName) => {
    setTheme(theme);
  };

  const handleToggleSidebar = () => {
    setToggleSidebar((prevState) => !prevState);
  };

  const handleChangeBackground = (background: string) => {
    setBackground(background);
  };

  return (
    <ThemeProvider value={theme}>
      <BackgroundFixed
        background={background}
        backgroundCollection={BACKGROUND_MAPPING_WITH_THUMBNAIL}
      />
      <ThemeSidebar
        isOpenSidebar={isOpenSidebar}
        handleToggleSidebar={handleToggleSidebar}
        themeVariantBox={
          <ThemeVariantList
            themeCollection={THEMES}
            selectingTheme={theme}
            handleChangeTheme={handleChangeTheme}
          />
        }
        backgroundVariantBox={
          <BackgroundVariantList
            backgroundCollection={BACKGROUND_THUMBNAIL}
            selectingBackground={background}
            handleChangeBackground={handleChangeBackground}
          />
        }
      />
      <div className="top-content d-flex flex-column flex-1 justify-content-between">
        <StoreProvider>
          <ErrorBoundary>
            <BrowserRouter>
              <Header handleToggleSidebar={handleToggleSidebar} />
              <Routes>
                <Route
                  path={ROUTES.ROOT}
                  element={<SpecBuilderPage />}
                />
                <Route
                  path={ROUTES.CART}
                  element={<CartPage />}
                />
                <Route
                  path={`${ROUTES.DETAIL}/:pcPartId`}
                  element={<PCPartDetail />}
                />
              </Routes>
            </BrowserRouter>
          </ErrorBoundary>
        </StoreProvider>
      </div>
      <div className="bottom-content">
        <Footer />
      </div>
    </ThemeProvider>
  );
};

export default App;
