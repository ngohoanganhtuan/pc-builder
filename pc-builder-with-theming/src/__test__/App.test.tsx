import { useState } from 'react';
import renderer from 'react-test-renderer';
import App from './../App';

jest.mock('react', () => {
  return {
    ...jest.requireActual('react'),
    useState: jest.fn(),
  };
});

const useStateMock = useState as jest.Mock;

describe('Test [App] component', () => {
  jest.spyOn(console, 'error').mockImplementation(() => jest.fn());
  test('Component [App] should render when dark theme is true', () => {
    useStateMock.mockImplementation(() => [true]);
    const component = renderer.create(<App />).toJSON();
    expect(component).toMatchSnapshot();
  });

  test('Component [App] should render when dark theme is false', () => {
    useStateMock.mockImplementation(() => [false]);
    const component = renderer.create(<App />).toJSON();
    expect(component).toMatchSnapshot();
  });
});
