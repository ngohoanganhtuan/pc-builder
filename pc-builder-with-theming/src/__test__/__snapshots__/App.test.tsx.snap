// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test [App] component Component [App] should render when dark theme is false 1`] = `
<div
  className="App"
>
  <div
    className="d-flex flex-column min-vh-100 justify-content-between"
  >
    <div
      className="top-content d-flex flex-column flex-grow-1 flex-shrink-1 flex-basis-0"
    >
      <div
        className="modal d-block"
        role="dialog"
      >
        <div
          className="modal-dialog shadow"
          role="document"
        >
          <div
            className="modal-content border-0"
          >
            <div
              className="modal-header bg-danger text-white"
            >
              <h5
                className="modal-title"
              >
                Error 
                <i
                  className="bi bi-emoji-frown"
                />
              </h5>
            </div>
            <div
              className="modal-body"
            >
              <details>
                <summary>
                  <span
                    className="text-danger"
                  >
                    TypeError
                  </span>
                   : 
                  Cannot read properties of undefined (reading 'pathname')
                </summary>
                <p
                  className="p-3"
                >
                  <small>
                    TypeError: Cannot read properties of undefined (reading 'pathname')
    at pathname (C:\\Users\\ASNET\\Desktop\\Training\\react-training\\practice-two\\node_modules\\.pnpm\\packages\\react-router\\lib\\components.tsx:197:5)
    at renderWithHooks (C:\\Users\\ASNET\\Desktop\\Training\\react-training\\practice-two\\node_modules\\.pnpm\\react-test-renderer@17.0.2_react@17.0.2\\node_modules\\react-test-renderer\\cjs\\react-test-renderer.development.js:6016:18)
    at mountIndeterminateComponent (C:\\Users\\ASNET\\Desktop\\Training\\react-training\\practice-two\\node_modules\\.pnpm\\react-test-renderer@17.0.2_react@17.0.2\\node_modules\\react-test-renderer\\cjs\\react-test-renderer.development.js:8744:13)
    at beginWork (C:\\Users\\ASNET\\Desktop\\Training\\react-training\\practice-two\\node_modules\\.pnpm\\react-test-renderer@17.0.2_react@17.0.2\\node_modules\\react-test-renderer\\cjs\\react-test-renderer.development.js:9966:16)
    at performUnitOfWork (C:\\Users\\ASNET\\Desktop\\Training\\react-training\\practice-two\\node_modules\\.pnpm\\react-test-renderer@17.0.2_react@17.0.2\\node_modules\\react-test-renderer\\cjs\\react-test-renderer.development.js:13800:12)
    at workLoopSync (C:\\Users\\ASNET\\Desktop\\Training\\react-training\\practice-two\\node_modules\\.pnpm\\react-test-renderer@17.0.2_react@17.0.2\\node_modules\\react-test-renderer\\cjs\\react-test-renderer.development.js:13728:5)
    at renderRootSync (C:\\Users\\ASNET\\Desktop\\Training\\react-training\\practice-two\\node_modules\\.pnpm\\react-test-renderer@17.0.2_react@17.0.2\\node_modules\\react-test-renderer\\cjs\\react-test-renderer.development.js:13691:7)
    at performSyncWorkOnRoot (C:\\Users\\ASNET\\Desktop\\Training\\react-training\\practice-two\\node_modules\\.pnpm\\react-test-renderer@17.0.2_react@17.0.2\\node_modules\\react-test-renderer\\cjs\\react-test-renderer.development.js:13404:18)
    at C:\\Users\\ASNET\\Desktop\\Training\\react-training\\practice-two\\node_modules\\.pnpm\\react-test-renderer@17.0.2_react@17.0.2\\node_modules\\react-test-renderer\\cjs\\react-test-renderer.development.js:2537:26
    at unstable_runWithPriority (C:\\Users\\ASNET\\Desktop\\Training\\react-training\\practice-two\\node_modules\\.pnpm\\scheduler@0.20.2\\node_modules\\scheduler\\cjs\\scheduler.development.js:468:12)
    at runWithPriority (C:\\Users\\ASNET\\Desktop\\Training\\react-training\\practice-two\\node_modules\\.pnpm\\react-test-renderer@17.0.2_react@17.0.2\\node_modules\\react-test-renderer\\cjs\\react-test-renderer.development.js:2486:10)
    at flushSyncCallbackQueueImpl (C:\\Users\\ASNET\\Desktop\\Training\\react-training\\practice-two\\node_modules\\.pnpm\\react-test-renderer@17.0.2_react@17.0.2\\node_modules\\react-test-renderer\\cjs\\react-test-renderer.development.js:2532:9)
    at flushSyncCallbackQueue (C:\\Users\\ASNET\\Desktop\\Training\\react-training\\practice-two\\node_modules\\.pnpm\\react-test-renderer@17.0.2_react@17.0.2\\node_modules\\react-test-renderer\\cjs\\react-test-renderer.development.js:2519:3)
    at scheduleUpdateOnFiber (C:\\Users\\ASNET\\Desktop\\Training\\react-training\\practice-two\\node_modules\\.pnpm\\react-test-renderer@17.0.2_react@17.0.2\\node_modules\\react-test-renderer\\cjs\\react-test-renderer.development.js:13004:9)
    at updateContainer (C:\\Users\\ASNET\\Desktop\\Training\\react-training\\practice-two\\node_modules\\.pnpm\\react-test-renderer@17.0.2_react@17.0.2\\node_modules\\react-test-renderer\\cjs\\react-test-renderer.development.js:16454:3)
    at Object.create (C:\\Users\\ASNET\\Desktop\\Training\\react-training\\practice-two\\node_modules\\.pnpm\\react-test-renderer@17.0.2_react@17.0.2\\node_modules\\react-test-renderer\\cjs\\react-test-renderer.development.js:17149:3)
    at Object.&lt;anonymous&gt; (C:\\Users\\ASNET\\Desktop\\Training\\react-training\\practice-two\\src\\__test__\\App.test.tsx:24:32)
    at Promise.then.completed (C:\\Users\\ASNET\\Desktop\\Training\\react-training\\practice-two\\node_modules\\.pnpm\\jest-circus@28.1.1\\node_modules\\jest-circus\\build\\utils.js:333:28)
    at new Promise (&lt;anonymous&gt;)
    at callAsyncCircusFn (C:\\Users\\ASNET\\Desktop\\Training\\react-training\\practice-two\\node_modules\\.pnpm\\jest-circus@28.1.1\\node_modules\\jest-circus\\build\\utils.js:259:10)
    at _callCircusTest (C:\\Users\\ASNET\\Desktop\\Training\\react-training\\practice-two\\node_modules\\.pnpm\\jest-circus@28.1.1\\node_modules\\jest-circus\\build\\run.js:277:40)
    at processTicksAndRejections (node:internal/process/task_queues:96:5)
    at _runTest (C:\\Users\\ASNET\\Desktop\\Training\\react-training\\practice-two\\node_modules\\.pnpm\\jest-circus@28.1.1\\node_modules\\jest-circus\\build\\run.js:209:3)
    at _runTestsForDescribeBlock (C:\\Users\\ASNET\\Desktop\\Training\\react-training\\practice-two\\node_modules\\.pnpm\\jest-circus@28.1.1\\node_modules\\jest-circus\\build\\run.js:97:9)
    at _runTestsForDescribeBlock (C:\\Users\\ASNET\\Desktop\\Training\\react-training\\practice-two\\node_modules\\.pnpm\\jest-circus@28.1.1\\node_modules\\jest-circus\\build\\run.js:91:9)
    at run (C:\\Users\\ASNET\\Desktop\\Training\\react-training\\practice-two\\node_modules\\.pnpm\\jest-circus@28.1.1\\node_modules\\jest-circus\\build\\run.js:31:3)
    at runAndTransformResultsToJestFormat (C:\\Users\\ASNET\\Desktop\\Training\\react-training\\practice-two\\node_modules\\.pnpm\\jest-circus@28.1.1\\node_modules\\jest-circus\\build\\legacy-code-todo-rewrite\\jestAdapterInit.js:135:21)
    at jestAdapter (C:\\Users\\ASNET\\Desktop\\Training\\react-training\\practice-two\\node_modules\\.pnpm\\jest-circus@28.1.1\\node_modules\\jest-circus\\build\\legacy-code-todo-rewrite\\jestAdapter.js:92:19)
    at runTestInternal (C:\\Users\\ASNET\\Desktop\\Training\\react-training\\practice-two\\node_modules\\.pnpm\\jest-runner@28.1.1\\node_modules\\jest-runner\\build\\runTest.js:411:16)
    at runTest (C:\\Users\\ASNET\\Desktop\\Training\\react-training\\practice-two\\node_modules\\.pnpm\\jest-runner@28.1.1\\node_modules\\jest-runner\\build\\runTest.js:499:34)
    at Object.worker (C:\\Users\\ASNET\\Desktop\\Training\\react-training\\practice-two\\node_modules\\.pnpm\\jest-runner@28.1.1\\node_modules\\jest-runner\\build\\testWorker.js:133:12)
                  </small>
                </p>
              </details>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      className="bottom-content"
    >
      <footer
        className="shadow-sm border bg-white footer"
      >
        <div
          className="container p-3 bg-indigo-500"
        >
          <h2
            className="fs-4 title"
          >
            PC Builder
          </h2>
          <p
            className="description"
          >
            PC Builder and the PC Builder logo are registered patents and trademarks of PC Builder. All other names, brands, products or services are trademarks or registered trademarks of their respective owners.
          </p>
          <div
            className="icon-list list-group"
          >
            <div
              className="d-flex flex-row gap-2"
            >
              <i
                className="bi bi-envelope icon"
              />
              <a
                className="text-decoration-none text-dark link"
                href="mailto:none"
              >
                pcbuilder@gmail.com
              </a>
            </div>
            <div
              className="d-flex flex-row gap-2"
            >
              <i
                className="bi bi-telephone icon"
              />
              <a
                className="text-decoration-none text-dark link"
                href="tel:none"
              >
                0236-0000-000
              </a>
            </div>
          </div>
        </div>
      </footer>
    </div>
  </div>
</div>
`;

exports[`Test [App] component Component [App] should render when dark theme is true 1`] = `
<div
  className="App dark"
>
  <div
    className="d-flex flex-column min-vh-100 justify-content-between"
  >
    <div
      className="top-content d-flex flex-column flex-grow-1 flex-shrink-1 flex-basis-0"
    >
      <div
        className="modal d-block"
        role="dialog"
      >
        <div
          className="modal-dialog shadow"
          role="document"
        >
          <div
            className="modal-content border-0"
          >
            <div
              className="modal-header bg-danger text-white"
            >
              <h5
                className="modal-title"
              >
                Error 
                <i
                  className="bi bi-emoji-frown"
                />
              </h5>
            </div>
            <div
              className="modal-body"
            >
              <details>
                <summary>
                  <span
                    className="text-danger"
                  >
                    TypeError
                  </span>
                   : 
                  Cannot read properties of undefined (reading 'pathname')
                </summary>
                <p
                  className="p-3"
                >
                  <small>
                    TypeError: Cannot read properties of undefined (reading 'pathname')
    at pathname (C:\\Users\\ASNET\\Desktop\\Training\\react-training\\practice-two\\node_modules\\.pnpm\\packages\\react-router\\lib\\components.tsx:197:5)
    at renderWithHooks (C:\\Users\\ASNET\\Desktop\\Training\\react-training\\practice-two\\node_modules\\.pnpm\\react-test-renderer@17.0.2_react@17.0.2\\node_modules\\react-test-renderer\\cjs\\react-test-renderer.development.js:6016:18)
    at mountIndeterminateComponent (C:\\Users\\ASNET\\Desktop\\Training\\react-training\\practice-two\\node_modules\\.pnpm\\react-test-renderer@17.0.2_react@17.0.2\\node_modules\\react-test-renderer\\cjs\\react-test-renderer.development.js:8744:13)
    at beginWork (C:\\Users\\ASNET\\Desktop\\Training\\react-training\\practice-two\\node_modules\\.pnpm\\react-test-renderer@17.0.2_react@17.0.2\\node_modules\\react-test-renderer\\cjs\\react-test-renderer.development.js:9966:16)
    at performUnitOfWork (C:\\Users\\ASNET\\Desktop\\Training\\react-training\\practice-two\\node_modules\\.pnpm\\react-test-renderer@17.0.2_react@17.0.2\\node_modules\\react-test-renderer\\cjs\\react-test-renderer.development.js:13800:12)
    at workLoopSync (C:\\Users\\ASNET\\Desktop\\Training\\react-training\\practice-two\\node_modules\\.pnpm\\react-test-renderer@17.0.2_react@17.0.2\\node_modules\\react-test-renderer\\cjs\\react-test-renderer.development.js:13728:5)
    at renderRootSync (C:\\Users\\ASNET\\Desktop\\Training\\react-training\\practice-two\\node_modules\\.pnpm\\react-test-renderer@17.0.2_react@17.0.2\\node_modules\\react-test-renderer\\cjs\\react-test-renderer.development.js:13691:7)
    at performSyncWorkOnRoot (C:\\Users\\ASNET\\Desktop\\Training\\react-training\\practice-two\\node_modules\\.pnpm\\react-test-renderer@17.0.2_react@17.0.2\\node_modules\\react-test-renderer\\cjs\\react-test-renderer.development.js:13404:18)
    at C:\\Users\\ASNET\\Desktop\\Training\\react-training\\practice-two\\node_modules\\.pnpm\\react-test-renderer@17.0.2_react@17.0.2\\node_modules\\react-test-renderer\\cjs\\react-test-renderer.development.js:2537:26
    at unstable_runWithPriority (C:\\Users\\ASNET\\Desktop\\Training\\react-training\\practice-two\\node_modules\\.pnpm\\scheduler@0.20.2\\node_modules\\scheduler\\cjs\\scheduler.development.js:468:12)
    at runWithPriority (C:\\Users\\ASNET\\Desktop\\Training\\react-training\\practice-two\\node_modules\\.pnpm\\react-test-renderer@17.0.2_react@17.0.2\\node_modules\\react-test-renderer\\cjs\\react-test-renderer.development.js:2486:10)
    at flushSyncCallbackQueueImpl (C:\\Users\\ASNET\\Desktop\\Training\\react-training\\practice-two\\node_modules\\.pnpm\\react-test-renderer@17.0.2_react@17.0.2\\node_modules\\react-test-renderer\\cjs\\react-test-renderer.development.js:2532:9)
    at flushSyncCallbackQueue (C:\\Users\\ASNET\\Desktop\\Training\\react-training\\practice-two\\node_modules\\.pnpm\\react-test-renderer@17.0.2_react@17.0.2\\node_modules\\react-test-renderer\\cjs\\react-test-renderer.development.js:2519:3)
    at scheduleUpdateOnFiber (C:\\Users\\ASNET\\Desktop\\Training\\react-training\\practice-two\\node_modules\\.pnpm\\react-test-renderer@17.0.2_react@17.0.2\\node_modules\\react-test-renderer\\cjs\\react-test-renderer.development.js:13004:9)
    at updateContainer (C:\\Users\\ASNET\\Desktop\\Training\\react-training\\practice-two\\node_modules\\.pnpm\\react-test-renderer@17.0.2_react@17.0.2\\node_modules\\react-test-renderer\\cjs\\react-test-renderer.development.js:16454:3)
    at Object.create (C:\\Users\\ASNET\\Desktop\\Training\\react-training\\practice-two\\node_modules\\.pnpm\\react-test-renderer@17.0.2_react@17.0.2\\node_modules\\react-test-renderer\\cjs\\react-test-renderer.development.js:17149:3)
    at Object.&lt;anonymous&gt; (C:\\Users\\ASNET\\Desktop\\Training\\react-training\\practice-two\\src\\__test__\\App.test.tsx:18:32)
    at Promise.then.completed (C:\\Users\\ASNET\\Desktop\\Training\\react-training\\practice-two\\node_modules\\.pnpm\\jest-circus@28.1.1\\node_modules\\jest-circus\\build\\utils.js:333:28)
    at new Promise (&lt;anonymous&gt;)
    at callAsyncCircusFn (C:\\Users\\ASNET\\Desktop\\Training\\react-training\\practice-two\\node_modules\\.pnpm\\jest-circus@28.1.1\\node_modules\\jest-circus\\build\\utils.js:259:10)
    at _callCircusTest (C:\\Users\\ASNET\\Desktop\\Training\\react-training\\practice-two\\node_modules\\.pnpm\\jest-circus@28.1.1\\node_modules\\jest-circus\\build\\run.js:277:40)
    at processTicksAndRejections (node:internal/process/task_queues:96:5)
    at _runTest (C:\\Users\\ASNET\\Desktop\\Training\\react-training\\practice-two\\node_modules\\.pnpm\\jest-circus@28.1.1\\node_modules\\jest-circus\\build\\run.js:209:3)
    at _runTestsForDescribeBlock (C:\\Users\\ASNET\\Desktop\\Training\\react-training\\practice-two\\node_modules\\.pnpm\\jest-circus@28.1.1\\node_modules\\jest-circus\\build\\run.js:97:9)
    at _runTestsForDescribeBlock (C:\\Users\\ASNET\\Desktop\\Training\\react-training\\practice-two\\node_modules\\.pnpm\\jest-circus@28.1.1\\node_modules\\jest-circus\\build\\run.js:91:9)
    at run (C:\\Users\\ASNET\\Desktop\\Training\\react-training\\practice-two\\node_modules\\.pnpm\\jest-circus@28.1.1\\node_modules\\jest-circus\\build\\run.js:31:3)
    at runAndTransformResultsToJestFormat (C:\\Users\\ASNET\\Desktop\\Training\\react-training\\practice-two\\node_modules\\.pnpm\\jest-circus@28.1.1\\node_modules\\jest-circus\\build\\legacy-code-todo-rewrite\\jestAdapterInit.js:135:21)
    at jestAdapter (C:\\Users\\ASNET\\Desktop\\Training\\react-training\\practice-two\\node_modules\\.pnpm\\jest-circus@28.1.1\\node_modules\\jest-circus\\build\\legacy-code-todo-rewrite\\jestAdapter.js:92:19)
    at runTestInternal (C:\\Users\\ASNET\\Desktop\\Training\\react-training\\practice-two\\node_modules\\.pnpm\\jest-runner@28.1.1\\node_modules\\jest-runner\\build\\runTest.js:411:16)
    at runTest (C:\\Users\\ASNET\\Desktop\\Training\\react-training\\practice-two\\node_modules\\.pnpm\\jest-runner@28.1.1\\node_modules\\jest-runner\\build\\runTest.js:499:34)
    at Object.worker (C:\\Users\\ASNET\\Desktop\\Training\\react-training\\practice-two\\node_modules\\.pnpm\\jest-runner@28.1.1\\node_modules\\jest-runner\\build\\testWorker.js:133:12)
                  </small>
                </p>
              </details>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      className="bottom-content"
    >
      <footer
        className="shadow-sm border bg-white footer"
      >
        <div
          className="container p-3 bg-indigo-500"
        >
          <h2
            className="fs-4 title"
          >
            PC Builder
          </h2>
          <p
            className="description"
          >
            PC Builder and the PC Builder logo are registered patents and trademarks of PC Builder. All other names, brands, products or services are trademarks or registered trademarks of their respective owners.
          </p>
          <div
            className="icon-list list-group"
          >
            <div
              className="d-flex flex-row gap-2"
            >
              <i
                className="bi bi-envelope icon"
              />
              <a
                className="text-decoration-none text-dark link"
                href="mailto:none"
              >
                pcbuilder@gmail.com
              </a>
            </div>
            <div
              className="d-flex flex-row gap-2"
            >
              <i
                className="bi bi-telephone icon"
              />
              <a
                className="text-decoration-none text-dark link"
                href="tel:none"
              >
                0236-0000-000
              </a>
            </div>
          </div>
        </div>
      </footer>
    </div>
  </div>
</div>
`;
